python_ta/transforms/type_inference_visitor.py

```diff
@@ -224,21 +224,54 @@ def _assign_type(self, target: NodeNG, expr_type: type, node: astroid.Assign) ->
         elif isinstance(target, astroid.Tuple):
             # Unpacking assignment, e.g. x, y = ...
             if isinstance(expr_type, typing.TupleMeta):
-                elt_inf_types = [self.lookup_inf_type(subtarget, subtarget.name) for subtarget in target.elts]
-                tuple_inf_type = wrap_container(Tuple, *elt_inf_types)
-                return self.type_constraints.unify(tuple_inf_type, expr_type, node)
+                assign_result = self._assign_tuple(target, expr_type, node)
             else:
-                iter_type_result = self._handle_call(target, '__iter__', expr_type)
+                assign_result = self._handle_call(target, '__iter__', expr_type)
                 for subtarget in target.elts:
-                    target_tvar = self.lookup_typevar(subtarget, subtarget.name)
-                    iter_type_result >> (
-                        lambda t: self.type_constraints.unify(target_tvar, t.__args__[0], node)
-                    )
-                return iter_type_result
+                    if isinstance(subtarget, astroid.Starred):
+                        target_tvar = self.lookup_typevar(subtarget.value, subtarget.value.name)
+                        unif_result = assign_result >> (
+                            lambda t: self.type_constraints.unify(target_tvar, List[t.__args__[0]], node))
+                    else:
+                        target_tvar = self.lookup_typevar(subtarget, subtarget.name)
+                        unif_result = assign_result >> (
+                            lambda t: self.type_constraints.unify(target_tvar, t.__args__[0], node))
+
+                    if isinstance(unif_result, TypeFail):
+                        return unif_result
+            return assign_result
         elif isinstance(target, astroid.Subscript):
             # TODO: previous case must recursively handle this one
             return self._handle_call(target, '__setitem__', target.value.inf_type, target.slice.inf_type, expr_type)
 
+    def _assign_tuple(self, target: astroid.Tuple, value: TupleMeta, node: astroid.Assign):
+        if any(isinstance(subtarget, astroid.Starred) for subtarget in target.elts):
```

Comment:
This method implementation should be simplified. Use the following approach:

1. Find the index of the `Starred` subtarget, if any. If there's more than one, stop. You can make another TypeFail subclass for this.
2. Using the index, partition the tuple into the types for the starred subtarget, and all other types. Similarly, extract a list of all non-starred subtargets.
3. Unify the non-starred subtargets with the "other types" from (2). This is currently lines 269-271. The same logic applies in all cases.
4. If there's a starred subtarget, unify it using the same approach you're currently using.